zetaclient/chains/solana/signer/withdraw_spl.go

```diff
@@ -0,0 +1,178 @@
+package signer
+
+import (
+	"context"
+
+	"cosmossdk.io/errors"
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/rpc"
+	"github.com/near/borsh-go"
+
+	"github.com/zeta-chain/node/pkg/chains"
+	contracts "github.com/zeta-chain/node/pkg/contracts/solana"
+	"github.com/zeta-chain/node/x/crosschain/types"
+)
+
+// createAndSignMsgWithdrawSPL creates and signs a withdraw spl message for gateway withdraw_spl instruction with TSS.
+func (signer *Signer) createAndSignMsgWithdrawSPL(
+	ctx context.Context,
+	params *types.OutboundParams,
+	height uint64,
+	asset string,
+	decimals uint8,
+	cancelTx bool,
+) (*contracts.MsgWithdrawSPL, error) {
+	chain := signer.Chain()
+	// #nosec G115 always positive
+	chainID := uint64(signer.Chain().ChainId)
+	nonce := params.TssNonce
+	amount := params.Amount.Uint64()
+
+	// zero out the amount if cancelTx is set. It's legal to withdraw 0 spl through the gateway.
+	if cancelTx {
+		amount = 0
+	}
+
+	// check receiver address
+	to, err := chains.DecodeSolanaWalletAddress(params.Receiver)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot decode receiver address %s", params.Receiver)
+	}
+
+	// parse token account
+	tokenAccount, err := solana.PublicKeyFromBase58(asset)
+	if err != nil {
+		return nil, err
+	}
+
+	// get recipient ata
+	recipientAta, _, err := solana.FindAssociatedTokenAddress(to, tokenAccount)
+	if err != nil {
+		return nil, err
+	}
+
+	// prepare withdraw spl msg and compute hash
+	msg := contracts.NewMsgWithdrawSPL(chainID, nonce, amount, decimals, tokenAccount, to, recipientAta)
+	msgHash := msg.Hash()
+
+	// sign the message with TSS to get an ECDSA signature.
+	// the produced signature is in the [R || S || V] format where V is 0 or 1.
+	signature, err := signer.TSS().Sign(ctx, msgHash[:], height, nonce, chain.ChainId, "")
+	if err != nil {
+		return nil, errors.Wrap(err, "Key-sign failed")
+	}
+	signer.Logger().Std.Info().Msgf("Key-sign succeed for chain %d nonce %d", chainID, nonce)
+
+	// attach the signature and return
+	return msg.SetSignature(signature), nil
+}
+
+// signWithdrawSPLTx wraps the withdraw spl 'msg' into a Solana transaction and signs it with the relayer key.
+func (signer *Signer) signWithdrawSPLTx(
+	ctx context.Context,
+	msg contracts.MsgWithdrawSPL,
+) (*solana.Transaction, error) {
+	// create withdraw spl instruction with program call data
+	var err error
+	var inst solana.GenericInstruction
```

Comment:
I wanted to suggest that we avoid using this pattern. It makes the code harder to understand and can lead to errors. It's also more complex and spreads the context across multiple functions.

Here's an example of what I mean:

```go
var foo SomeStruct

// ...
foo.a = 123

// ...

foo.b = "hello"

actMyWriteSomeSideEffects(&foo)

// ...

return foo, nil
```

Instead I propose the following:
```go
a := 123

// ...
b := "hello"

// ...

c,d := calculateSomething(...)

// ...

return SomeStruct{
  aa: a,
  bb: b,
  cc: c,
  dd: d,
}, nil
```